cibuildwheel/docker_container.py

```diff
@@ -155,12 +200,47 @@ def copy_out(self, from_path: PurePath, to_path: Path) -> None:
         # note: we assume from_path is a dir
         to_path.mkdir(parents=True, exist_ok=True)
 
-        subprocess.run(
-            f"docker exec -i {self.name} tar -cC {shell_quote(from_path)} -f - . | tar -xf -",
-            shell=True,
-            check=True,
-            cwd=to_path,
-        )
+        if self.container_engine == "podman":
+            # The copy out logic that works for docker does not seem to
+            # translate to podman, which seems to need the steps spelled out
+            # more explicitly.
+            command = f"{self.container_engine} exec -i {self.name} tar -cC {shell_quote(from_path)} -f /tmp/output-{self.name}.tar ."
+            subprocess.run(
+                command,
+                shell=True,
+                check=True,
+                cwd=to_path,
+                env=self.env,
+            )
+
+            command = f"{self.container_engine} cp {self.name}:/tmp/output-{self.name}.tar output-{self.name}.tar"
+            subprocess.run(
+                command,
+                shell=True,
+                check=True,
+                cwd=to_path,
+                env=self.env,
+            )
+            command = f"tar -xvf output-{self.name}.tar"
+            subprocess.run(
+                command,
+                shell=True,
+                check=True,
+                cwd=to_path,
+                env=self.env,
+            )
+            os.unlink(to_path / f"output-{self.name}.tar")
```

Comment:
I wonder if you tried a simple `podman cp` without the tar step?  In the Docker implementation, the tar step is to work around [a bug in Docker](https://github.com/moby/moby/issues/38995), but perhaps that bug doesn't exist in podman?

Reply:
It would make sense to at least document the alternative "simpler" route in case the bug is fixed in docker. I'll try it.

Reply:
I've added a stub for an alternate method: 

```python
        TRY_SIMPLE_CP = (self.container_engine == "podman")
        if TRY_SIMPLE_CP:
            # There is a bug in docker that prevents this simple implementation
            # from working https://github.com/moby/moby/issues/38995
            command = f"{self.container_engine} cp {self.name}:{shell_quote(from_path)} {shell_quote(to_path)}"
            subprocess.run(
                command,
                shell=True,
                check=True,
                cwd=to_path,
                env=self.env,
            )
```

But this causes the `test_dir_operations` test to fail. 

```
>           assert test_binary_data == (new_test_dir / "test.dat").read_bytes()

container  = <cibuildwheel.docker_container.DockerContainer object at 0x7f822933a640>
container_kwargs = {'container_engine': 'podman', 'docker_image': 'quay.io/pypa/manylinux2014_x86_64:2020-05-17-2f8ac3b', 'env': {'BASH_F...'/home/joncrall/.local/cuda::/home/joncrall/.local', 'CODE_DIR': '/home/joncrall/code', 'COLORTERM': 'truecolor', ...}}
dst_dir    = PurePosixPath('/tmp/test_dir')
dst_file   = PurePosixPath('/tmp/test_dir/test.dat')
new_test_dir = PosixPath('/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test_dir_new')
original_test_file = PosixPath('/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test.dat')
output     = '8\udcea\udc91\x1f\udcfd\udcf9X\udcb8H\udcb4\udce3\x06\udca7?\udc82\udcfd\x16\udcbd\x068\udcb6\x02\udca5\udcea\x0b{\x0...dc88ya\udcc9\x04\udc9a\udcb6\udc8d@ot\x08s\udc91\udcf7\udc88\udcb4,\udc82\x03\x17*6\udcba\x12[4w=\udcc4Bn)\udc9f\udce0'
test_binary_data = b'8\xea\x91\x1f\xfd\xf9X\xb8H\xb4\xe3\x06\xa7?\x82\xfd\x16\xbd\x068\xb6\x02\xa5\xea\x0b{\x03G\xda\xb0\xea\xa0\xdf\x93\...\x91h\xd2,\xbd\x1bd\xb3\r"\x88ya\xc9\x04\x9a\xb6\x8d@ot\x08s\x91\xf7\x88\xb4,\x82\x03\x17*6\xba\x12[4w=\xc4Bn)\x9f\xe0'
test_dir   = PosixPath('/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test_dir')
test_file  = PosixPath('/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test_dir/test.dat')
tmp_path   = PosixPath('/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1')

unit_test/docker_container_test.py:353: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../.pyenv/versions/3.9.9/lib/python3.9/pathlib.py:1259: in read_bytes
    with self.open(mode='rb') as f:
        self       = PosixPath('/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test_dir_new/test.dat')
../../.pyenv/versions/3.9.9/lib/python3.9/pathlib.py:1252: in open
    return io.open(self, mode, buffering, encoding, errors, newline,
        buffering  = -1
        encoding   = None
        errors     = None
        mode       = 'rb'
        newline    = None
        self       = PosixPath('/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test_dir_new/test.dat')
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = PosixPath('/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test_dir_new/test.dat')
name = '/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test_dir_new/test.dat', flags = 524288, mode = 438

    def _opener(self, name, flags, mode=0o666):
        # A stub for the opener argument to built-in open()
>       return self._accessor.open(self, flags, mode)
E       FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test_dir_new/test.dat'

flags      = 524288
mode       = 438
name       = '/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test_dir_new/test.dat'
self       = PosixPath('/tmp/pytest-of-joncrall/pytest-18/test_dir_operations_container_1/test_dir_new/test.dat')

../../.pyenv/versions/3.9.9/lib/python3.9/pathlib.py:1120: FileNotFoundError
============================================================================= short test summary info ==============================================================================
FAILED unit_test/docker_container_test.py::test_dir_operations[container_kwargs0] - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-joncrall/pytest-18/te...
FAILED unit_test/docker_container_test.py::test_dir_operations[container_kwargs1] - FileNotFoundError: [Errno 2] No such file or directory: '/tmp/pytest-of-joncrall/pytest-18/te...
```

Reply:
I have this in as a note in the code itself. I can leave it in as a node, (it is effectively commented out with an `if 0`) or just remove it. It seems like a simple podman cp won't work. Is there any major objection to the way I originally wrote it?

Reply:
Got a local podman setup working! I managed to get the simple cp to work, there's just a trick in how the src/dest paths are supplied on the command line (a bit like rsync). I've pushed that change.